### ISSUE TITLE
Have a single means to reference Resources

### ISSUE THREAD

**Issue by bevans2000** [mars-sim/mars-sim#306]
**Describe the bug/issue**
There are 2 way used to reference the same Resources. Some code uses Integer values from `ResourceUtil` whilst some use `ResourceAmount`.  The `Inventory` class for example has duplicated methods for both approaches but hold only a single value for each resource no matter how referenced.

** Proposal **
Having a single approach, probably `ResourceAmount`, would reduce the amount of duplicate code and provide a consistent approach.

**Comment by mokun** [comment]
@bevans2000 

I'm not familiar with `ResourceAmount` at all. 

We added `ResourceUtil` in the last 2 years as a utility class just for holding sets and maps of `AmountResource`.

Similarly, we added `ItemResourceUtil` for `ItemResource` or `Part`.

> The Inventory class for example has duplicated methods 

That's because we were transitioning into using the id of a `AmountResource` or `Part` for lookup.

The old way was keeping track of all the object references but we were experimenting with the use of the id.

Also, I suppose tracking the id of an object would bring the `Inventory` class one step closer to the Entity Framework but I could be wrong. 

It could also also be better for the future distributed platform as each node would need to send messages with the id of a `AmountResource` or  `Part` embedded in it to other nodes.

Q: should we develop our own in-house Entity Framework ?

**Comment by bevans2000** [comment]
My mistyping. I meant AmountResource. Having to use integer  identifiers in an object oriented language has to carefully considered.

**Comment by Urwumpe** [comment]
> Q: should we develop our own in-house Entity Framework ?

Short: I say no. 

First of all, I am not sure, if we need one and if we are too far into development already to describe things by a EF.

Next, if we need one and can support one, we should use one of the existing onces, which already are great and of a much better quality as we would reach in the next years of development.

So, no. Maybe it makes sense to use an existing entity framework. Could save us a lot of trouble in some fields, but would require a massive refactoring deep inside the core of MSP to implement. Again, my personal gut feeling there is: No, the work would be more effort than useful. 

But I did not do any math or estimates, so feel free to wack my head there and tell me how much I am wrong, its really a poor standard I set there.

**Comment by bevans2000** [comment]
I agreed these is no reason for a massive changes. Let's just standardise on referring to resources as AmountResource.

**Comment by bevans2000** [comment]
@mokun Should we include this for release 3.2.0 ? I find having 2 way to reference Resources is very confusing and creates extra code to maintain, e.g. switching between the 2 approaches.
This could be done via a gradual approach since `Inventory` appears to have duplicate methods for both forms of referencing.

**Comment by mokun** [comment]
> I find having 2 way to reference Resources is very confusing

It really depends on the situation.

Sometimes it's better to call a method asking for the resource id when the id is already known.

Say we have,

```
retrieveResource(AmountResource r) { }

retrieveResource(int resourceID) { }
```

So if I know the resourceID but don't know its AmountResource instance, then in that case, I'll call `retrieveResource(int resourceID)` and don't have to find the object instance in order to call that method `retrieveResource(AmountResource r)`.

But say if I already have the AmountResource instance, but don't know the resourceID, then r`etrieveResource(AmountResource r)`.

**Comment by mokun** [comment]
> Should we include this for release 3.2.0 ?

Let's postpone this to after 3.2.0 ? since it can break things. 😂

**Comment by bevans2000** [comment]
Ok After 3.2. We probably should call it 3.3 as it would be a feature change not a patch.
it would be a big change I agree

**Comment by bevans2000** [comment]
Currently Inventory can reference the contents through multiple means, either integer, `ItemResource` or `AmountResource`. The problem is further complicated by the integer range being partitioned to support Part/Item or Amount. This approach creates duplicate methods and the caller has to know what type of content is being added to the Inventory.
An option is to create a single class to describe the different type of content an Inventory can hold, e.g. `Commodity`. This would be a basic abstract class that contains a composite unique key based on the type of Commodity & an identifier. It could also contain things like base mass to make the mass calculations simpler.  Vary likely it could because the supercalss of the Part or AmountResource classes. The inventory would then have a single set of methods to add and remove Commodities. Note Units (Person, Vehicle, Robot) would still be handled separately.
If adopted the transition will not be easy but in the first phase replace Item/Part and then 2nd phase AmountResources.
This lay a good foundation to later tackle the large memory footprint of Inventory and provide specialised implementations for different Unit types.

**Comment by mokun** [comment]
> An option is to create a single class to describe the different type of content an Inventory can hold, e.g. Commodity.

ic.

> The inventory would then have a single set of methods to add and remove Commodities. Note Units (Person, Vehicle, Robot) would still be handled separately.

that would be cool.

> This lay a good foundation to later tackle the large memory footprint of Inventory and provide specialised implementations for different Unit types.

what's the current memory footprint of an inventory ? 

wonder if we can create some sort of benchmark unit test to show how each improvement would reduce the memory footprint, the speed of execution or the cpu utilization, etc.

**Comment by bevans2000** [comment]
This should be dropped from 3.4.0 as it is a massive undertaking. We need to choose either integers or the Part/AmountResource approach. Which ever way some code is going to change.

I also think we need to correct how we create the Integers because they are not constant and could change between a save & reload if the config was changed.

**Comment by mokun** [comment]
> because they are not constant and could change between a save & reload if the config was changed.

Thanks for pointing that out. Which one can be different when doing a reload ? 

That being the case, we have to be intentional with how we enumerate them and assign them an integer.

It needs to stay consistent.

**Comment by mokun** [comment]
Btw, so far, we NEVER provide any support to players who want to load from a previous version of saved sim., which makes life easier for us for now.

But if we one day successfully transition to JSON serialization, then it opens up a brand new new territory for us to look into every saved sim to see what's inside.

**Comment by bevans2000** [comment]
Yes good point about the configurations being stable between loads.

My issue is that there are many places throughout the code where we compare the int of a Resource to work out if it is an Equipment, Amount or Part. This is not good.

A good design would e to create a new ```ResourceID``` class that holds a single int value. Within the class we can then define predicates ```isPart```, ``isEquipment``` etc can put the selection logic in a single place. We could also extend this to have lookup methods embedded, ```getAmountResource``` or ```getPart``` and this would use the appropriate util class to resolve. 

It would make the code cleaner BUT when doing the change it will break a lot of places. So it needs to be a scheduled effort on a branch.

**Comment by mokun** [comment]
Are you planning to work on unifying how to reference a resource on 3.9.0 ?

**Comment by bevans2000** [comment]
I want to improve the map display and get back to correcting the transfer logic first

**Comment by mokun** [comment]
@bevans2000, 

Still want this to be worked in 3.9.0 ?

**Comment by bevans2000** [comment]
No, this is too big to add to the release

**Comment by bevans2000** [comment]
Let's still keep the Issue but not this release
